src/gadgets/mpt_update.rs

```diff
@@ -543,6 +377,293 @@ fn configure_storage<F: FieldExt>(cb: &mut ConstraintBuilder<F>, config: &MptUpd
+    );
+    cb.add_constraint(
+        "depth increased by 1",
+        config.selector.current(),
+        config.depth.delta() - Query::one(),
+    );
+    cb.add_constraint(
+        "0",
+        config
+            .segment_type
+            .previous_matches(SegmentType::Start)
+            .or(config
+                .segment_type
+                .previous_matches(SegmentType::AccountTrie)),
+        Query::one(),
+    );
+}
+
+fn configure_account_leaf0<F: FieldExt>(
+    cb: &mut ConstraintBuilder<F>,
+    config: &MptUpdateConfig,
+    segment_proof: &impl SegmentProofLookup,
+) {
+    cb.assert(
+        "from Start or AccountTrie",
+        config.selector.current(),
+        config
+            .segment_type
+            .previous_matches(SegmentType::Start)
+            .or(config
+                .segment_type
+                .previous_matches(SegmentType::AccountTrie)),
+    );
+    cb.assert(
+        "next is AccountLeaf1",
+        config.selector.current(),
+        config.segment_type.next_matches(SegmentType::AccountLeaf1),
+    );
+    cb.assert(
+        "path_type is Common",
+        config.selector.current(),
+        config.path_type.matches(PathType::Common),
+    );
+    cb.add_constraint(
+        "depth is 0",
+        config.selector.current(),
+        config.depth.current(),
+    );
+    cb.add_lookup(
+        "direction is correct for segment and proof types",
+        [
+            config.segment_type.current(),
+            config.proof_type.current(),
+            config.direction.current(),
+        ],
+        segment_proof.lookup(),
+    );
+
+    /* TODO: replaced by above SegmentProofLookup.
+
+    for variant in MPTProofType::iter() {
+        let conditional_constraints = |cb: &mut ConstraintBuilder<F>| match variant {
+            MPTProofType::AccountDestructed | MPTProofType::AccountDoesNotExist => todo!(),
+            MPTProofType::BalanceChanged
+            | MPTProofType::CodeHashExists
+            | MPTProofType::CodeSizeExists
+            | MPTProofType::NonceChanged
+            | MPTProofType::PoseidonCodeHashExists
+            | MPTProofType::StorageChanged
+            | MPTProofType::StorageDoesNotExist => cb.add_constraint(
+                "direction is 0",
+                config.selector.current(),
+                config.direction.current(),
+            ),
+        };
+        cb.condition(config.proof_type.matches(variant), conditional_constraints);
+    }
+    */
+
+    // add constraints that sibling = old_path_key and new_path_key
+}
+
+fn configure_account_leaf1<F: FieldExt>(
+    cb: &mut ConstraintBuilder<F>,
+    config: &MptUpdateConfig,
+    segment_proof: &impl SegmentProofLookup,
+) {
+    cb.assert(
+        "previous is AccountLeaf0",
+        config.selector.current(),
+        config
+            .segment_type
+            .previous_matches(SegmentType::AccountLeaf0),
+    );
+    cb.assert(
```

Comment:
Not sure if this gate (also, all gates for `account_leaf<x>`) is universal since we do not need more child for some proof type like poseidon hash (it laid in leaf1). In this case, `start` would be the next row in `segment_type` . For many of these conditions a lookup for the row transition in `segment_type` against proof type maybe more convenient

Reply:
Add `ProofSegmentPathLookup` for `(proof_type, previous_segement_type, current_segment_type)` lookup in commit
[7572019](https://github.com/scroll-tech/mpt-circuit/pull/32/commits/7572019d6ab2d913d6a09dc0a151fec94c4b04e3).